src/tools/clippy/clippy_utils/src/higher.rs

```diff
@@ -1,173 +1,210 @@
-        hir::BinOpKind::Rem => ast::BinOpKind::Rem,
-        hir::BinOpKind::Shl => ast::BinOpKind::Shl,
-        hir::BinOpKind::Shr => ast::BinOpKind::Shr,
-        hir::BinOpKind::Sub => ast::BinOpKind::Sub,
-    }
+/// The essential nodes of a desugared for loop as well as the entire span:
+/// `for pat in arg { body }` becomes `(pat, arg, body)`. Return `(pat, arg, body, span)`.
+pub struct ForLoop<'tcx> {
+    pub pat: &'tcx hir::Pat<'tcx>,
+    pub arg: &'tcx hir::Expr<'tcx>,
+    pub body: &'tcx hir::Expr<'tcx>,
+    pub span: Span
 }
 
-/// Represent a range akin to `ast::ExprKind::Range`.
-#[derive(Debug, Copy, Clone)]
-pub struct Range<'a> {
-    /// The lower bound of the range, or `None` for ranges such as `..X`.
-    pub start: Option<&'a hir::Expr<'a>>,
-    /// The upper bound of the range, or `None` for ranges such as `X..`.
-    pub end: Option<&'a hir::Expr<'a>>,
-    /// Whether the interval is open or closed.
-    pub limits: ast::RangeLimits,
+impl<'tcx> ForLoop<'tcx> {
+    #[inline]
+    pub fn hir(expr: &Expr<'tcx>) -> Option<Self> {
+        if_chain! {
+            if let hir::ExprKind::Match(ref iterexpr, ref arms, hir::MatchSource::ForLoopDesugar) = expr.kind;
+            if let Some(first_arm) = arms.get(0);
+            if let hir::ExprKind::Call(_, ref iterargs) = iterexpr.kind;
+            if let Some(first_arg) = iterargs.get(0);
+            if iterargs.len() == 1 && arms.len() == 1 && first_arm.guard.is_none();
+            if let hir::ExprKind::Loop(ref block, ..) = first_arm.body.kind;
+            if block.expr.is_none();
+            if let [ _, _, ref let_stmt, ref body ] = *block.stmts;
+            if let hir::StmtKind::Local(ref local) = let_stmt.kind;
+            if let hir::StmtKind::Expr(ref body_expr) = body.kind;
+            then {
+                return Some(Self {
+                    pat: &*local.pat,
+                    arg: first_arg,
+                    body: body_expr,
+                    span: first_arm.span
+                });
+            }
+        }
+        None
+    }
 }
 
-/// Higher a `hir` range to something similar to `ast::ExprKind::Range`.
-pub fn range<'a>(expr: &'a hir::Expr<'_>) -> Option<Range<'a>> {
-    /// Finds the field named `name` in the field. Always return `Some` for
-    /// convenience.
-    fn get_field<'c>(name: &str, fields: &'c [hir::ExprField<'_>]) -> Option<&'c hir::Expr<'c>> {
-        let expr = &fields.iter().find(|field| field.ident.name.as_str() == name)?.expr;
+pub struct If<'hir> {
+    pub cond: &'hir Expr<'hir>,
+    pub r#else: Option<&'hir Expr<'hir>>,
+    pub then: &'hir Expr<'hir>,
+}
 
-        Some(expr)
+impl<'hir> If<'hir> {
+    #[inline]
+    pub const fn hir(expr: &Expr<'hir>) -> Option<Self> {
+        if let ExprKind::If(
+            Expr { kind: ExprKind::DropTemps(cond), .. },
+            then,
+            r#else,
+        ) = expr.kind {
+            Some(Self { cond, r#else, then })
+        }
+        else {
```

Comment:
```suggestion
        } else {
```

Reply:
Is it possible to run rustfmt while developing Rustc? `./x.py fmt` doesn't touch `src/tools/clippy`

Reply:
Not 100% sure, but try to `cd` into the clippy dir and then run `cargo dev fmt`.

Reply:
`cargo fmt` did something but `cargo dev fmt` gives
```
error[internal]: left behind trailing whitespace
     --> ...rust/library/core/src/../../stdarch/crates/core_arch/src/mips/msa.rs:17403:17403:25
```